File: src/c++/src/epp/ext/EppExtension.cpp
/*******************************************************************************
 * The MIT License (MIT)
 *  
 * Copyright (c)2015 Neustar Inc.
 *  
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *  
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *  
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *******************************************************************************/
#include "EppExtension.hpp"
#include "EppUtil.hpp"
#include "EppCommandCreateSecDns.hpp"
#include "EppCommandUpdateSecDns.hpp"
#include "EppResponseDataInfoSecDns.hpp"
#include "EppCommandCreateLaunchRegistration.hpp"
#include "EppResponseDataCreateLaunchRegistration.hpp"
#include "EppCommandUpdateLaunchRegistration.hpp"
#include "EppCommandDeleteLaunchRegistration.hpp"
#include "EppCommandInfoLaunchRegistration.hpp"
#include "EppResponseDataInfoLaunchRegistration.hpp"
#include "EppCommandCheckLaunchRegistration.hpp"

#include "EppCommandCheckFee.hpp"
#include "EppCommandInfoFee.hpp"
#include "EppCommandCreateFee.hpp"
#include "EppCommandRenewFee.hpp"
#include "EppCommandUpdateFee.hpp"
#include "EppCommandTransferFee.hpp"

#include "EppResponseCheckFee.hpp"
#include "EppResponseInfoFee.hpp"
#include "EppResponseCreateFee.hpp"
#include "EppResponseRenewFee.hpp"
#include "EppResponseUpdateFee.hpp"
#include "EppResponseTransferFee.hpp"
#include "EppResponseDeleteFee.hpp"

#include "EppIDNData.hpp"
#include "EppResponseDataCheckClaims.hpp"
#include "EppAllocationTokenData.hpp"
#include "EppChangePollExt.hpp"

EppExtension::~EppExtension()
{
}

EppExtension * EppExtension::fromXML( const DOMNode &root )
{
	//if( root == null )
	//{
	//	return null;
	//}

	DOMString prefix = root.getPrefix();
	DOMString name = root.getLocalName();
	if( name.isNull() )
	{
		name = root.getNodeName();
	}
	if( name.isNull() )
	{
		return null;
	}

	EppExtension * entity = null;

	if(    ((prefix.isNull()) && name.equals("secDNS:create"))
	    || ((prefix.isNotNull()) && prefix.equals("secDNS") && name.equals("create")) )
	{
		unsigned short major, minor;
		major = minor = 0;
		entity = EppCommandCreateSecDns::fromXML(root);
		if( 0 == EppUtil::getDnsSecMajorMinor( root, major, minor) )
		{
			entity->secDnsVersion = setDnsSecVersion( major, minor );
		}
	}
	else if(    ((prefix.isNull()) && name.equals("secDNS:update"))
		 || ((prefix.isNotNull()) && prefix.equals("secDNS") && name.equals("update")) )
	{
		unsigned short  major, minor;
		major = minor = 0;
		entity = EppCommandUpdateSecDns::fromXML(root);
		if( 0 == EppUtil::getDnsSecMajorMinor( root, major, minor) )
		{
			entity->secDnsVersion = setDnsSecVersion( major, minor );
		}
	}
	else if(    ((prefix.isNull()) && name.equals("secDNS:infData"))
		|| ((prefix.isNotNull()) && prefix.equals("secDNS") && name.equals("infData")) )
	{
		unsigned short  major, minor;
		major = minor = 0;
		entity = EppResponseDataInfoSecDns::fromXML(root);
		if( 0 == EppUtil::getDnsSecMajorMinor( root, major, minor) )
		{
			entity->secDnsVersion = setDnsSecVersion( major, minor );
		}
	}
	else if(    ((prefix.isNull()) && name.equals("launch:create"))
		|| ((prefix.isNotNull()) && prefix.equals("launch") && name.equals("create")) )
	{
		entity = EppCommandCreateLaunchRegistration::fromXML(root);
	}
	else if(    ((prefix.isNull()) && name.equals("launch:creData"))
		|| ((prefix.isNotNull()) && prefix.equals("launch") && name.equals("creData")) )
	{
		entity = EppResponseDataCreateLaunchRegistration::fromXML(root);
	}
	else if(    ((prefix.isNull()) && name.equals("launch:update"))
		|| ((prefix.isNotNull()) && prefix.equals("launch") && name.equals("update")) )
	{
		entity = EppCommandUpdateLaunchRegistration::fromXML(root);
	}
	else if(    ((prefix.isNull()) && name.equals("launch:delete"))
		|| ((prefix.isNotNull()) && prefix.equals("launch") && name.equals("delete")) )
	{
		entity = EppCommandDeleteLaunchRegistration::fromXML(root);
	}
	else if(    ((prefix.isNull()) && name.equals("launch:info"))
		|| ((prefix.isNotNull()) && prefix.equals("launch") && name.equals("info")) )
	{
		entity = EppCommandInfoLaunchRegistration::fromXML(root);
	}
	else if(    ((prefix.isNull()) && name.equals("launch:infData"))
		|| ((prefix.isNotNull()) && prefix.equals("launch") && name.equals("infData")) )
	{
		entity = EppResponseDataInfoLaunchRegistration::fromXML(root);
	}
	else if(    ((prefix.isNull()) && name.equals("launch:check"))
		|| ((prefix.isNotNull()) && prefix.equals("launch") && name.equals("check")) )
	{
		entity = EppCommandCheckLaunchRegistration::fromXML(root); // TODO: One day we would need to pass correct ns name here
	}
	else if ( ((prefix.isNull()) && name.equals("idn:data"))
		|| ((prefix.isNotNull()) && prefix.equals("idn") && name.equals("data")) )
	{
		entity = EppIDNData::fromXML(root);
	}
	else if(    ((prefix.isNull()) && name.equals("launch:chkData"))
		|| ((prefix.isNotNull()) && prefix.equals("launch") && name.equals("chkData")) )
	{
		entity = EppResponseDataCheckClaims::fromXML(root);
	}
	else if(    ((prefix.isNull()) && name.equals("allocationToken:allocationToken"))
		|| ((prefix.isNotNull()) && prefix.equals("allocationToken") && name.equals("allocationToken")) )
	{
		entity = EppAllocationTokenData::fromXML(root);
	}
	else if(    ((prefix.isNull()) && name.equals("allocationToken:info"))
		|| ((prefix.isNotNull()) && prefix.equals("allocationToken") && name.equals("info")) )
	{
		entity = EppAllocationTokenData::fromXML(root);
	}

	//----------------Fee Extsnsion--------------------------
	else if(    (prefix.isNull() && name.equals("fee:check")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("check")) )
	{
		entity = EppCommandCheckFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:chkData")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("chkData")) )
	{
		entity = EppResponseCheckFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:info")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("info")) )
	{
		entity = EppCommandInfoFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:infData")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("infData")) )
	{
		entity = EppResponseInfoFee::fromXML(root);
	}
	// Nothing for transfer query request
	else if(    (prefix.isNull() && name.equals("fee:create")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("create")) )
	{
		entity = EppCommandCreateFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:creData")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("creData")) )
	{
		entity = EppResponseCreateFee::fromXML(root);
	}
	// nothing in delete request
	else if(    (prefix.isNull() && name.equals("fee:renew")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("renew")) )
	{
		entity = EppCommandRenewFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:renData")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("renData")) )
	{
		entity = EppResponseRenewFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:transfer")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("transfer")) )
	{
		entity = EppCommandTransferFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:trnData")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("trnData")) )
	{
		entity = EppResponseTransferFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:update")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("update")) )
	{
		entity = EppCommandUpdateFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:updData")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("updData")) )
	{
		entity = EppResponseUpdateFee::fromXML(root);
	}
	else if(    (prefix.isNull() && name.equals("fee:delData")) ||
		(prefix.isNotNull() && prefix.equals("fee") && name.equals("delData")) )
	{
		entity = EppResponseDeleteFee::fromXML(root);
	}
	//----------------Fee Extsnsion-end----------------------
        else if(    (prefix.isNull() && name.equals("changePoll:changeData")) ||
                (prefix.isNotNull() && prefix.equals("changePoll") && name.equals("changeData")) )
        {
                entity = EppChangePollExt::fromXML(root);
        }

	return entity;
}

File: src/java/src/com/neulevel/epp/ext/EppExtension.java
/*******************************************************************************
 * The MIT License (MIT)
 *  
 * Copyright (c) 2015 Neustar Inc.
 *  
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *  
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *  
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *******************************************************************************/
package com.neulevel.epp.ext;

import org.w3c.dom.Node;

import com.neulevel.epp.core.EppEntity;
import com.neulevel.epp.core.EppUtil;
import com.neulevel.epp.ext.command.EppCommandCheckLaunchRegistration;
import com.neulevel.epp.ext.command.EppCommandCreateLaunchRegistration;
import com.neulevel.epp.ext.command.EppCommandCreateSecDns;
import com.neulevel.epp.ext.command.EppCommandDeleteLaunchRegistration;
import com.neulevel.epp.ext.command.EppCommandInfoLaunchRegistration;
import com.neulevel.epp.ext.command.EppCommandUpdateLaunchRegistration;
import com.neulevel.epp.ext.command.EppCommandUpdateSecDns;
import com.neulevel.epp.ext.fee.command.EppCommandCheckFee;
import com.neulevel.epp.ext.fee.command.EppCommandCreateFee;
import com.neulevel.epp.ext.fee.command.EppCommandInfoFee;
import com.neulevel.epp.ext.fee.command.EppCommandRenewFee;
import com.neulevel.epp.ext.fee.command.EppCommandTransferFee;
import com.neulevel.epp.ext.fee.response.EppResponseCheckFee;
import com.neulevel.epp.ext.fee.response.EppResponseCreateFee;
import com.neulevel.epp.ext.fee.response.EppResponseInfoFee;
import com.neulevel.epp.ext.fee.response.EppResponseRenewFee;
import com.neulevel.epp.ext.fee.response.EppResponseTransferFee;
import com.neulevel.epp.ext.response.EppResponseDataCheckClaims;
import com.neulevel.epp.ext.response.EppResponseDataCreateLaunchRegistration;
import com.neulevel.epp.ext.response.EppResponseDataInfoLaunchRegistration;
import com.neulevel.epp.ext.response.EppResponseDataInfoSecDns;

/**
 * This <code>EppExtension</code> class implements EPP Extension entity.
 * 
 * @author Ning Zhang ning.zhang@neustar.com
 * @version $Revision: 1.9 $ $Date: 2013/07/08 21:36:29 $
 */
abstract public class EppExtension extends EppEntity
{
	/**
	 * Converts an XML element into an <code>EppExtension</code> object. The
	 * caller of this method must make sure that the root node is a child node
	 * of an EPP extension tag
	 * 
	 * @param root
	 *            root node for an <code>EppExtension</code> object in XML
	 *            format
	 * 
	 * @return an <code>EppEntity</code> object, or null if the node is invalid
	 */
	public static EppEntity fromXML(Node root)
	{
		if (root == null)
		{
			return null;
		}

		String prefix = root.getPrefix();
		String name = root.getLocalName();

		if ((prefix == null) || (name == null))
		{
			return null;
		}

		EppExtension entity = null;

		if (prefix.equals("secDNS"))
		{
			if (name.equals("create"))
			{
				entity = (EppExtension) EppCommandCreateSecDns.fromXML(root);
				if (null != entity)
				{
					String maj_min[] = EppUtil.getDnsSecMajorMinor(root);
					if (null != maj_min)
					{
						entity.secDnsMajor = maj_min[0];
						entity.secDnsMinor = maj_min[1];
					}
				}
			} else if (name.equals("update"))
			{
				entity = (EppExtension) EppCommandUpdateSecDns.fromXML(root);
				if (null != entity)
				{
					String maj_min[] = EppUtil.getDnsSecMajorMinor(root);
					if (null != maj_min)
					{
						entity.secDnsMajor = maj_min[0];
						entity.secDnsMinor = maj_min[1];
					}
				}
			} else if (name.equals("infData"))
			{
				entity = (EppExtension) EppResponseDataInfoSecDns.fromXML(root);
				if (null != entity)
				{
					String maj_min[] = EppUtil.getDnsSecMajorMinor(root);
					if (null != maj_min)
					{
						entity.secDnsMajor = maj_min[0];
						entity.secDnsMinor = maj_min[1];
					}
				}
			}
		} else if (prefix.equals("launch"))
		{
			if (name.equals("check"))
			{
				entity = (EppExtension) EppCommandCheckLaunchRegistration
						.fromXML(root);
			} else if (name.equals("create"))
			{
				entity = (EppExtension) EppCommandCreateLaunchRegistration
						.fromXML(root);
			} else if (name.equals("update"))
			{
				entity = (EppExtension) EppCommandUpdateLaunchRegistration
						.fromXML(root);
			} else if (name.equals("delete"))
			{
				entity = (EppExtension) EppCommandDeleteLaunchRegistration
						.fromXML(root);
			} else if (name.equals("info"))
			{
				entity = (EppExtension) EppCommandInfoLaunchRegistration
						.fromXML(root);
			} else if (name.equals("infData"))
			{
				entity = (EppExtension) EppResponseDataInfoLaunchRegistration
						.fromXML(root);
			} else if (name.equals("creData"))
			{
				entity = (EppExtension) EppResponseDataCreateLaunchRegistration
						.fromXML(root);
			} else if (name.equals("chkData"))
			{
				entity = (EppExtension) EppResponseDataCheckClaims
						.fromXML(root);
			}
		} else if (prefix.equals("idn"))
		{
			if (name.equals("data"))
			{
				entity = (EppExtension) EppIDNData.fromXML(root);
			}
		} else if (prefix.equals("allocationToken"))
		{
			if (name.equals("allocationToken"))
			{
				entity = (EppExtension) EppAllocationToken.fromXML(root);
			}
		} else if (prefix.equals("fee"))
		{
			if (name.equals("info")) {
				entity = (EppExtension) EppCommandInfoFee.fromXML(root);
			} else if (name.equals("infData")) {
				entity = (EppExtension) EppResponseInfoFee.fromXML(root);
			} else if (name.equals("check")) {
				entity = (EppExtension) EppCommandCheckFee.fromXML(root);
			} else if (name.equals("chkData")){
				entity = (EppExtension) EppResponseCheckFee.fromXML(root);
			} else if (name.equals("create")) {
				entity = (EppExtension) EppCommandCreateFee.fromXML(root);
			} else if (name.equals("creData")) {
				entity = (EppExtension) EppResponseCreateFee.fromXML(root);
			} else if (name.equals("transfer")) {
				entity = (EppExtension) EppCommandTransferFee.fromXML(root);
			} else if (name.equals("trnData")) {
				entity = (EppExtension) EppResponseTransferFee.fromXML(root);
			} else if (name.equals("renew")) {
				entity = (EppExtension) EppCommandRenewFee.fromXML(root);
			} else if (name.equals("renData")) {
				entity = (EppExtension) EppResponseRenewFee.fromXML(root);
			}
		} else if (prefix.equals("changePoll") && name.equals("changeData") ) 
		{
			entity = (EppExtension) EppChangePollExt.fromXML(root);
			
		}
		return (EppEntity) entity;
	}

	protected String secDnsMajor = "1";
	protected String secDnsMinor = "1";
}
